chirp/drivers/baofeng_digital.py

```diff
@@ -0,0 +1,492 @@
+        0x08: ('Chinese', 'Spanish'),
+        0x09: ('English', 'Spanish'),
+        0x0A: ('Chinese', 'German'),
+        0x0B: ('English', 'German'),
+        0x0C: ('Chinese', 'French'),
+        0x0D: ('English', 'French'),
+    }
+
+    @property
+    def _memsize(self):
+        return self._proto.mem_chunks[-1].end + 1
+
+    def get_features(self):
+        rf = chirp_common.RadioFeatures()
+        rf.has_rx_dtcs = True
+        rf.has_bank = False
+        rf.has_name = False
+        rf.has_tuning_step = False
+        rf.can_odd_split = True
+        rf.has_cross = True
+        rf.memory_bounds = (1, self._proto.chan_count)
+        rf.valid_bands = self.VALID_BANDS
+        rf.valid_duplexes = ["", "+", "-", "split"]
+        rf.valid_modes = ['FM', 'NFM', 'DIG', 'DN']
+        rf.valid_power_levels = self.POWER_LEVELS
+        rf.valid_tuning_steps = [2.5, 6.25, 5.0]
+        rf.valid_tmodes = ["", "Tone", "TSQL", "DTCS", "Cross"]
+        rf.valid_cross_modes = [
+            "Tone->Tone",
+            "Tone->DTCS",
+            "DTCS->Tone",
+            "DTCS->",
+            "->Tone",
+            "->DTCS",
+            "DTCS->DTCS",
+        ]
+        return rf
+
+    def _get_radio_id(self):
+        for attempt in range(5):
+            try:
+                baofeng_common._clean_buffer(self)
+
+                self.pipe.write(b'\x02prOGRAM')
+                resp = self.pipe.read(4)
+                if len(resp) != 4:
+                    raise errors.RadioError(
+                        'Radio did not send identification')
+                return resp
+            except errors.RadioError:
+                # Don't wait if already last attempt
+                if attempt == 4:
+                    raise
+                self._exit_programming()
+                time.sleep(1)
+
+    def _enter_programming(self, write):
+        radio_id = self._get_radio_id()
+        if radio_id[1] != self._serial_id:
+            raise errors.RadioError('Detected a different radio model')
+
+        self._metadata['fw_region'] = radio_id[0]
+        self._metadata['fw_version'] = radio_id[2] * 0x100 + radio_id[3]
+
+        # Issue protocol-specific command
+        self.pipe.write(b'\x02' + self._proto.serial_magic + b'OGRAM')
+        if self.pipe.read(1) != CMD_ACK:
+            raise errors.RadioError('No ACK on protocol-specific command')
+
+        # Issue fixed commands and check their values
+        if write:
+            preamble = self._PREAMBLE_CMDS + self._WRITE_PREAMBLE_CMDS
+        else:
+            preamble = self._PREAMBLE_CMDS
+        for step, cmd in enumerate(preamble, 1):
+            self.pipe.write(cmd[0])
+            if self.pipe.read(len(cmd[1])) != cmd[1]:
+                raise errors.RadioError(
+                    f'Radio replied incorrectly to preamble step {step}')
+
+    def _clean_buffer(self):
+        self.pipe.timeout = 0.001
+        self.pipe.read(256)
+        self.pipe.timeout = 0.1
```

Comment:
Have you tested this on Windows? On windows, setting the timeout flushes the buffer, which probably doesn't matter here but it sometimes surprises people when they do it in the middle of a conversation with the radio.

Reply:
Yep, I'm using Windows in fact for all development. No issues.